pkg/virt-handler/node-labeller/node_labeller.go

```diff
@@ -0,0 +1,208 @@
+	configMapInformer cache.SharedIndexInformer
+	clientset         kubecli.KubevirtClient
+	Queue             workqueue.RateLimitingInterface
+	host              string
+	namespace         string
+}
+
+func NewNodeLabeller(kvmController *device_manager.DeviceController, configMapInformer cache.SharedIndexInformer, clientset kubecli.KubevirtClient, host, namespace string) *NodeLabeller {
+	return &NodeLabeller{
+		kvmController:     kvmController,
+		configMapInformer: configMapInformer,
+		clientset:         clientset,
+		Queue:             workqueue.NewRateLimitingQueue(workqueue.DefaultControllerRateLimiter()),
+
+		host:      host,
+		namespace: namespace,
+	}
+}
+
+//Run runs node-labeller
+func (n *NodeLabeller) Run(threadiness int, stop chan struct{}) {
+	defer n.Queue.ShutDown()
+	logger := log.DefaultLogger()
+	logger.Infof("node-labeller is running")
+
+	n.configMapInformer.AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc: func(obj interface{}) {
+			n.Queue.AddRateLimited(obj)
+		},
+		DeleteFunc: func(obj interface{}) {
+			n.Queue.AddRateLimited(obj)
+		},
+		UpdateFunc: func(old, new interface{}) {
+			n.Queue.AddRateLimited(new)
+		},
+	})
+
+	go n.configMapInformer.Run(stop)
+	cache.WaitForCacheSync(stop, n.configMapInformer.HasSynced)
+
+	for i := 0; i < threadiness; i++ {
+		go wait.Until(n.runWorker, time.Second, stop)
+	}
+	<-stop
+}
+func (n *NodeLabeller) runWorker() {
+	for n.Execute() {
+	}
+}
+
+func (n *NodeLabeller) Execute() bool {
+	key, quit := n.Queue.Get()
+	if quit {
+		return false
+	}
+	defer n.Queue.Done(key)
+	err := n.run()
+
+	if err != nil {
+		n.Queue.AddRateLimited(key)
+	} else {
+		n.Queue.Forget(key)
+	}
+	return true
+}
+
+func (n *NodeLabeller) run() error {
+	logger := log.DefaultLogger()
+
+	if !n.kvmController.NodeHasDevice(device_manager.KVMPath) {
+		return goerror.New(fmt.Sprintf("Node-labeller cannot work without KVM device."))
+	}
+
+	cpuFeatures := make(map[string]bool)
+	cpuModels := make([]string, 0)
+	//parse all informations
+	cpuModels, cpuFeatures, err := n.getCPUInfo()
+	if err != nil {
+		logger.Infof("node-labeller cannot get new labels %s\n", err.Error())
+		return err
+	}
+	//get node on which virt-handler is running
+	node, err := n.clientset.CoreV1().Nodes().Get(n.host, metav1.GetOptions{})
+	if err != nil {
+		logger.Infof("node-labeller cannot get node %s", n.host)
+		return err
+	}
+
+	//prepare new labels
+	newLabels := n.prepareLabels(cpuModels, cpuFeatures)
+	//remove old labeller labels
+	n.removeCPULabels(node, n.getNodeLabellerLabels(node))
+	//add new labels
+	n.addNodeLabels(node, newLabels)
+	//patch node with new labels
+	data := []byte(fmt.Sprintf(`[{"op": "replace", "path": "/metadata/labels", "value": {%s}}, {"op": "replace", "path": "/metadata/annotations", "value": {%s}}]`, convertMapToText(node.Labels), convertMapToText(node.Annotations)))
+	_, err = n.clientset.CoreV1().Nodes().Patch(node.Name, types.JSONPatchType, data)
```

Comment:
we should be able to do a `reflect.DeepEqual()` on the old and new labels in order to know of a Patch is required or not. If we calculated the exact same labels as what's already on the node, then no Patch would be required